FocusOn/ViewModel/ProgressViewModel.swift

```diff
@@ -0,0 +1,125 @@
+//
+//  ProgressViewModel.swift
+//  FocusOn
+//
+//  Created by Alexandra Ivanova on 12/07/2023.
+//
+
+import Foundation
+import SwiftUICharts
+import SwiftUI
+
+class ProgressViewModel: ObservableObject {
+    @Published var allGoals: [Goal]?
+    private let dataService: DataServiceProtocol
+    let calendar = Calendar.current
+    let firstWeekday = 2
+    
+    let noGoal = Legend(color: .gray, label: "No goal", order: 1)
+    let fail = Legend(color: .red, label: "Fail", order: 2)
+    let smallProgress = Legend(color: .orange, label: "Small Progress", order: 3)
+    let bigProgress = Legend(color: .yellow, label: "Big Progress", order: 4)
+    let success = Legend(color: .green, label: "Success", order: 5)
+    
+    init( dataService: DataServiceProtocol = MockDataService()) {
+        self.dataService = dataService
+        self.allGoals = dataService.allGoals
+    }
+    
+    func fillChartData() -> [[DataPoint]]? {
+        guard let unwrappedGoals = allGoals else {
+            return nil
+        }
+        let weeklyChunksOfDataPoints = generateDataPointsForWeeklyChunksOf(goals: unwrappedGoals)
+        
+        return weeklyChunksOfDataPoints.isEmpty ? nil : weeklyChunksOfDataPoints
+    }
+    
+    private func calculateNumberOfCompletedTasks(goal: Goal) -> Int {
+        var numberOfCompletedTasks = 0
+        
+        for task in goal.tasks where task.isCompleted {
+            numberOfCompletedTasks += 1
+        }
+        return numberOfCompletedTasks
+    }
+    
+    private func calculateGoalProgress(goal: Goal) -> Legend {
+        let legendSignature: Legend
+        
+        switch calculateNumberOfCompletedTasks(goal: goal) {
+        case 0:
+            legendSignature = fail
+        case 1:
+            legendSignature = smallProgress
+        case 2:
+            legendSignature = bigProgress
+        case 3:
+            legendSignature = success
+        default:
+            legendSignature = noGoal
+        }
+        return legendSignature
+    }
+    
+    private func generateDataPointForGoal(goal: Goal) -> DataPoint {
+        let barHeight = Double(calculateNumberOfCompletedTasks(goal: goal) + 1)
+        let dayComponent = Calendar.current.component(.day, from: goal.createdAt)
+        let label = LocalizedStringKey(String(dayComponent))
+        let legend = calculateGoalProgress(goal: goal)
+        
+        return DataPoint(value: barHeight, label: label, legend: legend)
+    }
+    
+    private func splitGoalsIntoWeeklyChunks(goals: [Goal]) -> [[Goal]] {
```

Comment:
I think this method would be much simpler if you used the calendar component `weekOfYear` and `year`.

The idea  of the method is to quite simply regroup the data that belongs to the same week together.